src/evolve.test.ts

```diff
@@ -0,0 +1,465 @@
+import { pipe } from './pipe';
+import { evolve } from './evolve';
+import { omit } from './omit';
+import { set } from './set';
+import { map } from './map';
+import { add } from './add';
+
+describe('data first', () => {
+  it('creates a new object by evolving the `object` according to the `transformation` functions', function () {
+    const transf = {
+      count: add(1),
+      data: { elapsed: add(1), remaining: add(-1) },
+    };
+    const object = {
+      id: 10,
+      count: 10,
+      data: { elapsed: 100, remaining: 1400 },
+    };
+    const expected = {
+      id: 10,
+      count: 11,
+      data: { elapsed: 101, remaining: 1399 },
+    };
+    const result = evolve(object, transf);
+    expect(result).toEqual(expected);
+    expectTypeOf(result).toEqualTypeOf<typeof expected>();
+  });
+
+  it('does not invoke function if object does not contain the key', function () {
+    const transf = { n: add(1), m: add(1) };
+    const object = { m: 3 };
+    const expected = { m: 4 };
+    const result = evolve(object, transf);
+    expect(result).toEqual(expected);
+    expectTypeOf(result).toEqualTypeOf<typeof expected>();
+  });
+
+  it('is not destructive and is immutable', function () {
+    const transf = { n: add(1) };
+    const object = { n: 100 };
+    const expected = { n: 101 };
+    const result = evolve(object, transf);
+    expect(object).toEqual({ n: 100 });
+    expect(result).toEqual(expected);
+    expect(result).not.toBe(expected);
+  });
+
+  it('is recursive', function () {
+    const transf = { nested: { second: add(-1), third: add(1) } };
+    const object = { first: 1, nested: { second: 2, third: 3 } };
+    const expected = { first: 1, nested: { second: 1, third: 4 } };
+    const result = evolve(object, transf);
+    expect(result).toEqual(expected);
+    expectTypeOf(result).toEqualTypeOf<typeof expected>();
+  });
+
+  it('ignores primitive value transformations', function () {
+    const transf = { n: 2, m: 'foo' };
+    const object = { n: 0, m: 1 };
+    const expected = { n: 0, m: 1 };
+    const result = evolve(object, transf);
+    expect(result).toEqual(expected);
+  });
+
+  it('ignores null transformations', function () {
+    const transf = { n: null };
+    const object = { n: 0 };
+    const expected = { n: 0 };
+    const result = evolve(object, transf);
+    expect(result).toEqual(expected);
+    expectTypeOf(result).toEqualTypeOf<typeof expected>();
+  });
+
+  it('creates a new array by evolving the `array` according to the `transformation` functions', function () {
+    // NOTE:
+    // If we use tuple in `transformations` parameter,
+    // use `as const` or `evolve` can't handle typing.
```

Comment:
We could stop handling functions in arrays, but for now, we are adapting to the Ramda implementation.